File: code/causal_inference/diff_in_diff.py
"""
Difference-in-Differences (DiD)
================================

Implementation of DiD estimation with various extensions.

References:
    Card & Krueger (1994), Bertrand et al. (2004), Abadie (2005)
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from sklearn.linear_model import LinearRegression
from typing import Tuple, Optional, List
from dataclasses import dataclass

sns.set_style("whitegrid")


@dataclass
class DiDResults:
    """Results from DiD estimation."""
    att: float  # Average treatment effect on the treated
    se: float
    ci_lower: float
    ci_upper: float
    n_treat: int
    n_control: int
    parallel_trends_pvalue: Optional[float] = None


class DifferenceInDifferences:
    """
    Difference-in-Differences estimator.

    Parameters
    ----------
    data : pd.DataFrame
        Panel data with columns: unit_id, time_period, outcome, treated
    outcome_col : str
        Name of outcome variable
    treated_col : str
        Name of treatment indicator
    time_col : str
        Name of time period variable
    unit_col : str
        Name of unit identifier
    """

    def __init__(self, data: pd.DataFrame, outcome_col: str = 'outcome',
                 treated_col: str = 'treated', time_col: str = 'post',
                 unit_col: str = 'unit_id'):
        self.data = data.copy()
        self.outcome_col = outcome_col
        self.treated_col = treated_col
        self.time_col = time_col
        self.unit_col = unit_col

    def estimate_2x2(self, cluster_se: bool = True) -> DiDResults:
        """
        Estimate 2x2 DiD (two time periods, two groups).

        Parameters
        ----------
        cluster_se : bool
            Use cluster-robust standard errors

        Returns
        -------
        DiDResults
            Estimation results
        """
        df = self.data

        # Calculate group-time means
        means = df.groupby([self.treated_col, self.time_col])[self.outcome_col].mean()

        # DiD estimator
        Y_11 = means.loc[(1, 1)]  # Treated, post
        Y_10 = means.loc[(1, 0)]  # Treated, pre
        Y_01 = means.loc[(0, 1)]  # Control, post
        Y_00 = means.loc[(0, 0)]  # Control, pre

        att = (Y_11 - Y_10) - (Y_01 - Y_00)

        # Regression approach for standard errors
        df['treat_post'] = df[self.treated_col] * df[self.time_col]

        # OLS: Y = β0 + β1*Treat + β2*Post + β3*Treat*Post + ε
        X = df[[self.treated_col, self.time_col, 'treat_post']].values
        X_with_const = np.column_stack([np.ones(len(df)), X])
        y = df[self.outcome_col].values

        beta = np.linalg.lstsq(X_with_const, y, rcond=None)[0]
        att_reg = beta[3]  # Interaction coefficient

        # Standard errors
        residuals = y - X_with_const @ beta

        if cluster_se:
            # Cluster-robust standard errors at unit level
            clusters = df[self.unit_col].values
            unique_clusters = np.unique(clusters)
            n_clusters = len(unique_clusters)

            # Calculate clustered variance
            meat = np.zeros((X_with_const.shape[1], X_with_const.shape[1]))
            for cluster in unique_clusters:
                idx = clusters == cluster
                X_cluster = X_with_const[idx]
                resid_cluster = residuals[idx]
                meat += np.outer(X_cluster.T @ resid_cluster, X_cluster.T @ resid_cluster)

            bread_inv = np.linalg.inv(X_with_const.T @ X_with_const / len(df))
            var_beta = bread_inv @ meat @ bread_inv * n_clusters / (n_clusters - 1)
        else:
            # Heteroskedasticity-robust standard errors
            meat = X_with_const.T @ (residuals[:, np.newaxis]**2 * X_with_const)
            bread_inv = np.linalg.inv(X_with_const.T @ X_with_const)
            var_beta = bread_inv @ meat @ bread_inv

        se = np.sqrt(var_beta[3, 3])

        # Confidence interval
        ci_lower = att - 1.96 * se
        ci_upper = att + 1.96 * se

        n_treat = df[df[self.treated_col] == 1][self.unit_col].nunique()
        n_control = df[df[self.treated_col] == 0][self.unit_col].nunique()

        return DiDResults(
            att=att,
            se=se,
            ci_lower=ci_lower,
            ci_upper=ci_upper,
            n_treat=n_treat,
            n_control=n_control
        )

    def test_parallel_trends(self, pre_periods: List[int]) -> float:
        """
        Test parallel trends assumption using pre-treatment periods.

        Parameters
        ----------
        pre_periods : list
            List of pre-treatment time periods

        Returns
        -------
        float
            p-value for parallel trends test
        """
        df_pre = self.data[self.data[self.time_col].isin(pre_periods)]

        # Regress outcome on time trends separately for treated and control
        results_treat = []
        results_control = []

        for period in pre_periods:
            treat_mean = df_pre[(df_pre[self.time_col] == period) &
                               (df_pre[self.treated_col] == 1)][self.outcome_col].mean()
            control_mean = df_pre[(df_pre[self.time_col] == period) &
                                 (df_pre[self.treated_col] == 0)][self.outcome_col].mean()
            results_treat.append(treat_mean)
            results_control.append(control_mean)

        # Test if trends are parallel (interaction of time with treatment)
        n_periods = len(pre_periods)
        X = np.column_stack([
            np.ones(n_periods * 2),
            np.repeat([0, 1], n_periods),  # Treatment indicator
            np.tile(pre_periods, 2),  # Time
            np.repeat([0, 1], n_periods) * np.tile(pre_periods, 2)  # Interaction
        ])

        y = np.concatenate([results_control, results_treat])

        beta = np.linalg.lstsq(X, y, rcond=None)[0]
        residuals = y - X @ beta

        # F-test for interaction coefficient
        sigma2 = np.sum(residuals**2) / (len(y) - X.shape[1])
        var_beta = sigma2 * np.linalg.inv(X.T @ X)
        t_stat = beta[3] / np.sqrt(var_beta[3, 3])
        p_value = 2 * (1 - stats.t.cdf(abs(t_stat), len(y) - X.shape[1]))

        return p_value

    def event_study(self, time_periods: List[int], base_period: int = -1) -> pd.DataFrame:
        """
        Estimate event study (dynamic DiD).

        Parameters
        ----------
        time_periods : list
            List of time periods relative to treatment
        base_period : int
            Base period (normalized to 0)

        Returns
        -------
        pd.DataFrame
            Event study coefficients
        """
        df = self.data.copy()

        # Create period indicators
        estimates = []

        for t in time_periods:
            if t == base_period:
                # Normalized period
                estimates.append({'period': t, 'coef': 0, 'se': 0, 'ci_lower': 0, 'ci_upper': 0})
                continue

            # Interaction between treatment and period indicator
            df[f'period_{t}'] = (df[self.time_col] == t).astype(int)
            df[f'treat_period_{t}'] = df[self.treated_col] * df[f'period_{t}']

            # Run regression
            period_cols = [c for c in df.columns if c.startswith('period_') or c.startswith('treat_period_')]
            X = df[[self.treated_col] + period_cols].values
            X_with_const = np.column_stack([np.ones(len(df)), X])
            y = df[self.outcome_col].values

            beta = np.linalg.lstsq(X_with_const, y, rcond=None)[0]

            # Find coefficient for this period's interaction
            col_idx = period_cols.index(f'treat_period_{t}') + 1 + 1  # +1 for const, +1 for treated main effect
            coef = beta[col_idx]

            # Standard error (simplified)
            residuals = y - X_with_const @ beta
            sigma2 = np.sum(residuals**2) / (len(y) - X_with_const.shape[1])
            var_beta = sigma2 * np.linalg.inv(X_with_const.T @ X_with_const)
            se = np.sqrt(var_beta[col_idx, col_idx])

            ci_lower = coef - 1.96 * se
            ci_upper = coef + 1.96 * se

            estimates.append({'period': t, 'coef': coef, 'se': se,
                            'ci_lower': ci_lower, 'ci_upper': ci_upper})

        return pd.DataFrame(estimates)

    def plot_did(self, results: DiDResults):
        """
        Visualize DiD design.

        Parameters
        ----------
        results : DiDResults
            Estimation results
        """
        df = self.data

        # Calculate group-time means
        means = df.groupby([self.treated_col, self.time_col])[self.outcome_col].mean().unstack()

        fig, axes = plt.subplots(1, 2, figsize=(14, 5))

        # Left panel: Parallel trends plot
        ax = axes[0]
        time_points = means.columns
        ax.plot(time_points, means.loc[0], 'o-', linewidth=2, markersize=8,
               label='Control', color='blue')
        ax.plot(time_points, means.loc[1], 's-', linewidth=2, markersize=8,
               label='Treated', color='red')

        # Counterfactual (parallel trends)
        if len(time_points) == 2:
            control_change = means.loc[0, time_points[1]] - means.loc[0, time_points[0]]
            counterfactual = means.loc[1, time_points[0]] + control_change
            ax.plot([time_points[0], time_points[1]],
                   [means.loc[1, time_points[0]], counterfactual],
                   '--', linewidth=2, color='red', alpha=0.5,
                   label='Counterfactual (Treated)')

            # Show treatment effect
            ax.annotate('', xy=(time_points[1], means.loc[1, time_points[1]]),
                       xytext=(time_points[1], counterfactual),
                       arrowprops=dict(arrowstyle='<->', color='green', lw=2))
            mid_y = (means.loc[1, time_points[1]] + counterfactual) / 2
            ax.text(time_points[1] + 0.05, mid_y, f'ATT={results.att:.2f}',
                   fontsize=12, color='green')

        ax.set_xlabel('Time Period')
        ax.set_ylabel('Outcome')
        ax.set_title('Difference-in-Differences')
        ax.legend()
        ax.grid(True, alpha=0.3)

        # Right panel: Distribution of treatment effects
        ax = axes[1]

        # Individual-level treatment effects (for visualization)
        df_pre = df[df[self.time_col] == 0]
        df_post = df[df[self.time_col] == 1]

        df_merged = df_pre.merge(df_post, on=self.unit_col, suffixes=('_pre', '_post'))
        df_merged['diff'] = df_merged[f'{self.outcome_col}_post'] - df_merged[f'{self.outcome_col}_pre']

        treat_diffs = df_merged[df_merged[f'{self.treated_col}_post'] == 1]['diff']
        control_diffs = df_merged[df_merged[f'{self.treated_col}_post'] == 0]['diff']

        ax.hist(control_diffs, bins=20, alpha=0.5, label='Control', density=True, color='blue')
        ax.hist(treat_diffs, bins=20, alpha=0.5, label='Treated', density=True, color='red')
        ax.axvline(x=control_diffs.mean(), color='blue', linestyle='--', linewidth=2)
        ax.axvline(x=treat_diffs.mean(), color='red', linestyle='--', linewidth=2)
        ax.set_xlabel('Change in Outcome')
        ax.set_ylabel('Density')
        ax.set_title('Distribution of Changes')
        ax.legend()

        plt.tight_layout()
        return fig


def simulate_did_data(n_units: int = 500, n_periods: int = 2,
                     treatment_effect: float = 5.0) -> pd.DataFrame:
    """
    Simulate panel data for DiD.

    Parameters
    ----------
    n_units : int
        Number of units
    n_periods : int
        Number of time periods
    treatment_effect : float
        True ATT

    Returns
    -------
    pd.DataFrame
        Simulated panel data
    """
    np.random.seed(42)

    # Unit fixed effects
    unit_fe = np.random.normal(10, 2, n_units)

    # Treatment assignment (some units treated)
    treated = np.random.binomial(1, 0.5, n_units)

    # Create panel
    data = []
    for t in range(n_periods):
        post = int(t >= n_periods // 2)

        for i in range(n_units):
            # Outcome with unit FE, time trend, and treatment effect
            y = (unit_fe[i] +
                 3 * t +  # Common time trend
                 treatment_effect * treated[i] * post +  # Treatment effect
                 np.random.normal(0, 1))

            data.append({
                'unit_id': i,
                'time': t,
                'post': post,
                'treated': treated[i],
                'outcome': y
            })

    return pd.DataFrame(data)


def example_minimum_wage():
    """
    Example: Minimum wage effect on employment (Card & Krueger style).
    """
    print("Example: Minimum Wage and Employment")
    print("=" * 70)

    # Simulate data
    np.random.seed(789)
    n = 400

    # New Jersey (treated) vs Pennsylvania (control)
    state = np.random.choice(['NJ', 'PA'], n, p=[0.5, 0.5])
    treated = (state == 'NJ').astype(int)

    # Create panel data (before and after minimum wage increase)
    data = []
    for period, post in [(0, 0), (1, 1)]:
        for i in range(n):
            # Employment (FTE employees)
            baseline = 20 + np.random.normal(0, 5)
            time_effect = 2 * post
            treatment_effect = 3.0 * treated[i] * post  # Positive effect (contrary to theory!)

            employment = baseline + time_effect + treatment_effect + np.random.normal(0, 2)

            data.append({
                'unit_id': i,
                'state': state[i],
                'time': period,
                'post': post,
                'treated': treated[i],
                'outcome': employment
            })

    df = pd.DataFrame(data)

    # Estimate DiD
    did = DifferenceInDifferences(df)
    results = did.estimate_2x2(cluster_se=True)

    print(f"Estimated ATT: {results.att:.3f}")
    print(f"Standard error: {results.se:.3f}")
    print(f"95% CI: [{results.ci_lower:.3f}, {results.ci_upper:.3f}]")
    print(f"N (treated): {results.n_treat}")
    print(f"N (control): {results.n_control}")

    # Plot
    fig = did.plot_did(results)
    plt.savefig('did_minimum_wage.png', dpi=150, bbox_inches='tight')
    plt.show()


if __name__ == "__main__":
    example_minimum_wage()

File: 04-causal-inference/causal-inference-fundamentals/code/r/diff_in_diff.R
# Difference-in-Differences (DiD) in R
# =====================================
#
# Implementation using fixest for two-way fixed effects (TWFE)
#
# References:
#   Card & Krueger (1994), Bertrand et al. (2004)

# Load required packages
suppressPackageStartupMessages({
  library(fixest)   # For fast fixed effects estimation
  library(did)      # For Callaway-Sant'Anna estimator
  library(ggplot2)
  library(dplyr)
  library(tidyr)
  library(broom)
})

# ============================================
# Simulate DiD Data
# ============================================

simulate_did_data <- function(n_units = 500, n_periods = 10,
                             treatment_time = 5, treatment_effect = 5.0,
                             seed = 42) {
  #' Simulate panel data for DiD
  #'
  #' @param n_units Number of units
  #' @param n_periods Number of time periods
  #' @param treatment_time When treatment starts
  #' @param treatment_effect True ATT
  #' @return data.frame with panel data

  set.seed(seed)

  # Unit fixed effects
  unit_fe <- rnorm(n_units, 10, 2)

  # Treatment assignment (50% treated)
  treated_units <- sample(1:n_units, n_units/2)
  treated <- rep(0, n_units)
  treated[treated_units] <- 1

  # Create panel
  data <- expand.grid(
    unit_id = 1:n_units,
    time = 1:n_periods
  )

  data$treated <- treated[data$unit_id]
  data$post <- as.integer(data$time >= treatment_time)
  data$treat_post <- data$treated * data$post

  # Outcome with unit FE, time trends, and treatment effect
  data$outcome <- unit_fe[data$unit_id] +
                 3 * data$time +  # Common time trend
                 treatment_effect * data$treat_post +  # Treatment effect
                 rnorm(nrow(data), 0, 1)

  data
}

# ============================================
# 2x2 DiD Estimation
# ============================================

estimate_did_2x2 <- function(data, outcome = "outcome",
                             treated = "treated", post = "post",
                             unit_id = "unit_id") {
  #' Estimate 2x2 DiD (two periods, two groups)
  #'
  #' @param data data.frame with panel data
  #' @param outcome Name of outcome variable
  #' @param treated Name of treatment indicator
  #' @param post Name of post-treatment indicator
  #' @param unit_id Name of unit identifier
  #' @return List with results

  # Calculate group-time means
  means <- data %>%
    group_by(!!sym(treated), !!sym(post)) %>%
    summarise(mean_outcome = mean(!!sym(outcome)), .groups = "drop")

  # Extract means
  Y_11 <- means %>% filter(!!sym(treated) == 1, !!sym(post) == 1) %>%
          pull(mean_outcome)
  Y_10 <- means %>% filter(!!sym(treated) == 1, !!sym(post) == 0) %>%
          pull(mean_outcome)
  Y_01 <- means %>% filter(!!sym(treated) == 0, !!sym(post) == 1) %>%
          pull(mean_outcome)
  Y_00 <- means %>% filter(!!sym(treated) == 0, !!sym(post) == 0) %>%
          pull(mean_outcome)

  # DiD estimator
  att <- (Y_11 - Y_10) - (Y_01 - Y_00)

  # Regression approach for inference
  data$treat_post <- data[[treated]] * data[[post]]

  # With cluster-robust SE at unit level
  did_reg <- feols(
    as.formula(paste(outcome, "~", treated, "+", post, "+ treat_post")),
    data = data,
    cluster = unit_id
  )

  list(
    att = att,
    model = did_reg,
    means = means
  )
}

# ============================================
# Two-Way Fixed Effects (TWFE) DiD
# ============================================

estimate_twfe_did <- function(data, outcome = "outcome",
                              treat_post = "treat_post",
                              unit_id = "unit_id", time = "time") {
  #' Estimate DiD using two-way fixed effects
  #'
  #' @param data Panel data
  #' @param outcome Outcome variable
  #' @param treat_post Treatment x Post interaction
  #' @param unit_id Unit identifier
  #' @param time Time variable
  #' @return feols object

  formula_str <- paste(outcome, "~", treat_post, "| ", unit_id, "+", time)

  twfe_model <- feols(
    as.formula(formula_str),
    data = data,
    cluster = unit_id
  )

  twfe_model
}

# ============================================
# Event Study (Dynamic DiD)
# ============================================

estimate_event_study <- function(data, outcome = "outcome",
                                 treated = "treated", time = "time",
                                 treatment_time = 5, unit_id = "unit_id",
                                 pre_periods = 4, post_periods = 5) {
  #' Estimate event study / dynamic DiD
  #'
  #' @param data Panel data
  #' @param outcome Outcome variable
  #' @param treated Treatment indicator
  #' @param time Time variable
  #' @param treatment_time Time when treatment starts
  #' @param unit_id Unit identifier
  #' @param pre_periods Number of pre-treatment periods
  #' @param post_periods Number of post-treatment periods
  #' @return feols object

  # Create relative time variable
  data$rel_time <- data[[time]] - treatment_time
  data$rel_time[data[[treated]] == 0] <- 0  # Set to 0 for never-treated

  # Create event time dummies (excluding -1 as base period)
  event_times <- c(-pre_periods:(-2), 0:post_periods)

  for (t in event_times) {
    data[[paste0("treat_", t)]] <- as.integer(
      data[[treated]] == 1 & data$rel_time == t
    )
  }

  # Event study formula
  treat_vars <- paste0("treat_", event_times)
  formula_str <- paste(
    outcome, "~",
    paste(treat_vars, collapse = " + "),
    "|", unit_id, "+", time
  )

  event_model <- feols(
    as.formula(formula_str),
    data = data,
    cluster = unit_id
  )

  event_model
}

# ============================================
# Visualization Functions
# ============================================

plot_did_trends <- function(data, results, outcome = "outcome",
                           treated = "treated", time = "time") {
  #' Plot parallel trends
  #'
  #' @param data Panel data
  #' @param results DiD results object
  #' @param outcome Outcome variable
  #' @param treated Treatment indicator
  #' @param time Time variable

  # Calculate time-series means by group
  trends <- data %>%
    group_by(!!sym(time), !!sym(treated)) %>%
    summarise(mean_outcome = mean(!!sym(outcome)), .groups = "drop") %>%
    mutate(group = ifelse(!!sym(treated) == 1, "Treated", "Control"))

  p <- ggplot(trends, aes(x = !!sym(time), y = mean_outcome,
                         color = group, group = group)) +
    geom_line(linewidth = 1.2) +
    geom_point(size = 3) +
    geom_vline(xintercept = 5, linetype = "dashed", alpha = 0.5) +
    annotate("text", x = 5, y = max(trends$mean_outcome),
            label = "Treatment", vjust = -0.5) +
    scale_color_manual(values = c("Control" = "blue", "Treated" = "red")) +
    labs(
      x = "Time Period",
      y = "Average Outcome",
      title = "Difference-in-Differences: Parallel Trends",
      color = "Group"
    ) +
    theme_minimal() +
    theme(
      text = element_text(size = 12),
      legend.position = "bottom"
    )

  print(p)
  invisible(p)
}

plot_event_study <- function(event_model, pre_periods = 4, post_periods = 5) {
  #' Plot event study coefficients
  #'
  #' @param event_model Event study feols object
  #' @param pre_periods Number of pre-treatment periods
  #' @param post_periods Number of post-treatment periods

  # Extract coefficients
  coefs <- tidy(event_model, conf.int = TRUE)

  # Parse relative time from variable names
  coefs$rel_time <- as.numeric(gsub("treat_", "", coefs$term))

  # Add base period (-1) at zero
  base_period <- data.frame(
    term = "treat_-1",
    estimate = 0,
    std.error = 0,
    statistic = 0,
    p.value = 1,
    conf.low = 0,
    conf.high = 0,
    rel_time = -1
  )

  coefs <- bind_rows(coefs, base_period) %>%
    arrange(rel_time)

  # Plot
  p <- ggplot(coefs, aes(x = rel_time, y = estimate)) +
    geom_line(linewidth = 1) +
    geom_point(size = 3) +
    geom_errorbar(aes(ymin = conf.low, ymax = conf.high), width = 0.2) +
    geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
    geom_vline(xintercept = -0.5, linetype = "dotted", alpha = 0.5) +
    annotate("text", x = -0.5, y = max(coefs$conf.high),
            label = "Treatment", vjust = -0.5, size = 3) +
    labs(
      x = "Periods Relative to Treatment",
      y = "Estimated Effect",
      title = "Event Study Plot",
      subtitle = "Point estimates with 95% confidence intervals"
    ) +
    theme_minimal() +
    theme(text = element_text(size = 12))

  print(p)
  invisible(p)
}

# ============================================
# Example 1: Minimum Wage and Employment
# ============================================

example_minimum_wage <- function() {
  cat("Example 1: Minimum Wage and Employment (Card & Krueger)\n")
  cat("========================================================\n\n")

  set.seed(789)
  n <- 400

  # New Jersey (treated) vs Pennsylvania (control)
  state <- sample(c("NJ", "PA"), n, replace = TRUE, prob = c(0.5, 0.5))
  treated <- as.integer(state == "NJ")

  # Create panel (before and after minimum wage increase)
  data <- data.frame(
    unit_id = rep(1:n, 2),
    state = rep(state, 2),
    treated = rep(treated, 2),
    time = rep(c(0, 1), each = n)
  )

  data$post <- data$time

  # Employment (FTE employees)
  baseline <- rnorm(n, 20, 5)
  data$outcome <- rep(baseline, 2) +
                 2 * data$time +  # Time trend
                 3.0 * data$treated * data$post +  # Treatment effect
                 rnorm(nrow(data), 0, 2)

  # Estimate DiD
  cat("2x2 DiD Estimation:\n")
  cat("-" * 70, "\n")
  did_results <- estimate_did_2x2(data)

  cat(sprintf("DiD Estimate (ATT): %.3f\n", did_results$att))
  cat("\nRegression Results:\n")
  print(summary(did_results$model))

  # Plot
  p <- plot_did_trends(data, did_results)
  ggsave("did_minimum_wage_R.png", plot = p, width = 10, height = 6, dpi = 150)

  invisible(did_results)
}

# ============================================
# Example 2: Event Study
# ============================================

example_event_study <- function() {
  cat("\nExample 2: Event Study (Dynamic DiD)\n")
  cat("=====================================\n\n")

  # Simulate multi-period data
  data <- simulate_did_data(n_units = 300, n_periods = 10,
                           treatment_time = 5, treatment_effect = 5.0)

  # Estimate event study
  event_model <- estimate_event_study(data, treatment_time = 5,
                                     pre_periods = 4, post_periods = 5)

  cat("Event Study Results:\n")
  print(summary(event_model))

  # Plot
  p <- plot_event_study(event_model)
  ggsave("did_event_study_R.png", plot = p, width = 10, height = 6, dpi = 150)

  invisible(event_model)
}

# ============================================
# Example 3: TWFE with Multiple Treatment Times
# ============================================

example_staggered_adoption <- function() {
  cat("\nExample 3: Staggered Treatment Adoption\n")
  cat("========================================\n\n")

  set.seed(999)

  n_units <- 500
  n_periods <- 10

  # Staggered treatment adoption
  treatment_times <- sample(c(4, 6, 8, Inf), n_units, replace = TRUE,
                            prob = c(0.25, 0.25, 0.25, 0.25))

  # Create panel
  data <- expand.grid(
    unit_id = 1:n_units,
    time = 1:n_periods
  )

  data$treatment_time <- treatment_times[data$unit_id]
  data$treated <- as.integer(
    data$time >= data$treatment_time & is.finite(data$treatment_time)
  )

  # Outcome
  unit_fe <- rnorm(n_units, 10, 2)
  data$outcome <- unit_fe[data$unit_id] +
                 2 * data$time +
                 4.0 * data$treated +
                 rnorm(nrow(data), 0, 1)

  # TWFE estimation
  twfe_model <- feols(outcome ~ treated | unit_id + time,
                     data = data, cluster = ~ unit_id)

  cat("Two-Way Fixed Effects Results:\n")
  print(summary(twfe_model))

  cat("\nNote: With staggered treatment and heterogeneous effects,\n")
  cat("TWFE may give biased estimates. Consider Callaway-Sant'Anna\n")
  cat("or Sun-Abraham estimators instead.\n")

  invisible(twfe_model)
}

# ============================================
# Test Parallel Trends Assumption
# ============================================

test_parallel_trends <- function(data, outcome = "outcome",
                                treated = "treated", time = "time",
                                treatment_time = 5) {
  #' Test parallel trends in pre-treatment period
  #'
  #' @param data Panel data
  #' @param outcome Outcome variable
  #' @param treated Treatment indicator
  #' @param time Time variable
  #' @param treatment_time When treatment starts

  # Pre-treatment data only
  data_pre <- data %>% filter(!!sym(time) < treatment_time)

  # Test for differential trends
  data_pre$time_treat <- data_pre[[time]] * data_pre[[treated]]

  trend_test <- feols(
    as.formula(paste(outcome, "~ time * ", treated)),
    data = data_pre,
    cluster = ~ unit_id
  )

  cat("\n")
  cat("=" * 70, "\n")
  cat("Parallel Trends Test (Pre-Treatment Period)\n")
  cat("=" * 70, "\n")
  print(summary(trend_test))

  cat("\nIf the interaction coefficient (time:", treated, ") is insignificant,\n")
  cat("this supports the parallel trends assumption.\n")

  invisible(trend_test)
}

# ============================================
# Run all examples
# ============================================

if (sys.nframe() == 0) {
  cat("\n")
  cat("=" * 70, "\n")
  cat("Difference-in-Differences Examples in R\n")
  cat("=" * 70, "\n\n")

  example_minimum_wage()
  example_event_study()
  example_staggered_adoption()

  # Test parallel trends for simulated data
  data_test <- simulate_did_data(n_units = 300, n_periods = 10,
                                treatment_time = 5)
  test_parallel_trends(data_test, treatment_time = 5)

  cat("\nAll examples completed successfully!\n")
}
